client/matching/lb.go

```diff
@@ -0,0 +1,136 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"fmt"
+	"math/rand"
+	"strings"
+
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/service/dynamicconfig"
+)
+
+type (
+	// LoadBalancer is the interface for implementers of
+	// component that distributes add/poll api calls across
+	// available task list partitions when possible
+	LoadBalancer interface {
+		// PickWritePartition returns the task list partition for adding
+		// an activity or decision task. The input is the name of the
+		// original task list (with no partition info). When forwardedFrom
+		// is non-empty, this call is forwardedFrom from a child partition
+		// to a parent partition in which case, no load balancing should be
+		// performed
+		PickWritePartition(
+			domainID string,
+			taskList shared.TaskList,
+			taskListType int,
+			forwardedFrom string,
+		) string
+
+		// PickReadPartition returns the task list partition to send a poller to.
+		// Input is name of the original task list as specified by caller. When
+		// forwardeFrom is non-empty, no load balancing should be done.
+		PickReadPartition(
+			domainID string,
+			taskList shared.TaskList,
+			taskListType int,
+			forwardedFrom string,
+		) string
+	}
+
+	defaultLoadBalancer struct {
+		nReadPartitions  dynamicconfig.IntPropertyFnWithTaskListInfoFilters
+		nWritePartitions dynamicconfig.IntPropertyFnWithTaskListInfoFilters
+		domainIDToName   func(string) (string, error)
+	}
+)
+
+const (
+	taskListPartitionPrefix = "/__cadence_sys/"
+)
+
+// NewLoadBalancer returns an instance of matching load balancer that
+// can help distribute api calls across task list partitions
+func NewLoadBalancer(
+	domainIDToName func(string) (string, error),
+	dc *dynamicconfig.Collection,
+) LoadBalancer {
+	return &defaultLoadBalancer{
+		domainIDToName:   domainIDToName,
+		nReadPartitions:  dc.GetIntPropertyFilteredByTaskListInfo(dynamicconfig.MatchingNumTasklistReadPartitions, 1),
+		nWritePartitions: dc.GetIntPropertyFilteredByTaskListInfo(dynamicconfig.MatchingNumTasklistWritePartitions, 1),
+	}
+}
+
+func (lb *defaultLoadBalancer) PickWritePartition(
+	domainID string,
+	taskList shared.TaskList,
+	taskListType int,
+	forwardedFrom string,
+) string {
+	return lb.pickPartition(domainID, taskList, taskListType, forwardedFrom, lb.nWritePartitions)
+}
+
+func (lb *defaultLoadBalancer) PickReadPartition(
+	domainID string,
+	taskList shared.TaskList,
+	taskListType int,
+	forwardedFrom string,
+) string {
+	return lb.pickPartition(domainID, taskList, taskListType, forwardedFrom, lb.nReadPartitions)
+}
+
+func (lb *defaultLoadBalancer) pickPartition(
+	domainID string,
+	taskList shared.TaskList,
+	taskListType int,
+	forwardedFrom string,
+	nPartitions dynamicconfig.IntPropertyFnWithTaskListInfoFilters,
+) string {
+
+	if forwardedFrom != "" || taskList.GetKind() == shared.TaskListKindSticky {
```

Comment:
Nope, multiple levels of forwarding is ok. May be a comment might help here.  When forwardedFrom is set, LB doesn't pick a partition because partition is already picked by the caller